js/core/src/background-action.ts

```diff
@@ -0,0 +1,314 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { JSONSchema7 } from 'json-schema';
+import * as z from 'zod';
+import { Action, ActionMetadata, defineAction, Middleware } from './action.js';
+import { ActionContext } from './context.js';
+import { GenkitError } from './error.js';
+import { ActionType, Registry } from './registry.js';
+import { toJsonSchema } from './schema.js';
+
+/**
+ * Zod schema of an opration representing a background task.
+ */
+export const OperationSchema = z.object({
+  action: z.string().optional(),
+  id: z.string(),
+  done: z.boolean().optional(),
+  output: z.any().optional(),
+  metadata: z.record(z.string(), z.any()).optional(),
+});
+
+/**
+ * Background operation.
+ */
+export interface Operation<O = any> {
+  action?: string;
+  id: string;
+  done?: boolean;
+  output?: O;
+  metadata?: Record<string, any>;
+}
```

Comment:
What happens if the operation errors? I think we can choose between:

1. The `check` operation throws.
2. We add an `error?: {message: string, [key: string]: unknown}`.

I think (2) is probably the way to go -- which I think should be a small change to the existing code and just some additional code on the Veo integration side.

Reply:
I was kinda thinking #1, but agree #2 is better. Done.